### ISSUE TITLE
Force training

### ISSUE THREAD

**Issue by Dankomaister** [atomistic-machine-learning/schnetpack#5]
Was wondering what is the proper way to add forces to the loss function?
Currently I have my loss function defined like this

```
def loss(batch, result):
	
	N = torch.sum((batch['_atomic_numbers'] != 0).float(), 1, keepdim=True) # Nr of atoms per image
	
	gamma = (result['y']/N - batch['energy']/N)**2
	
	return torch.sum(gamma)
```
which works fine.
However if i try to add forces to the loss function like this

```
def loss(batch, result):
	
	N = torch.sum((batch['_atomic_numbers'] != 0).float(), 1, keepdim=True) # Nr of atoms per image
	
	gammaE = (result['y']/N - batch['energy']/N)**2
	gammaF = torch.sum(torch.sum((result['dydx'] - batch['forces'])**2, 2), 1, keepdim=True)/(3*N)
	
	return torch.sum(gammaE + gammaF)

```
it stops working. I get a value of the loss function for the first image then the function just returns `nan`. Not sure what the problem is here, is there a better way to include forces in the loss function? or is there a bug somewhere.

**Comment by mgastegger** [comment]
Hi, I tried both versions of the loss functions in the schnetpack_md17.py script and they seem to work fine on CPU and GPU with the models training properly.

Maybe this is an issue with the data.
Which dataset are you using?

Another possibility might be the pytorch version ('0.4.1.post2' on my machine).

**Comment by Dankomaister** [comment]
We are using our own dataset generated using VASP and saved as an ASE db file
[all_images.zip](https://github.com/atomistic-machine-learning/schnetpack/files/2393614/all_images.zip)
our version of pytorch is 0.5.0a0+a24163a I can try and use another dataset.

**Comment by mgastegger** [comment]
Hm, I have to check whether this could be a problem with the periodic boundary conditions and forces.

**Comment by Dankomaister** [comment]
Ok, tested with the md17 dataset and that is working. I will keep testing with our dataset, is schnetpack not built for periodic boundary conditions?

**Comment by mgastegger** [comment]
Periodic boundary conditions should work, as is the case for energies. This is most likely a problem of the combination of PBC and forces. I'm currently running a few tests.

**Comment by mgastegger** [comment]
There was a problem with the forces, where the derivative of the square root in the distance computation lead to zero divisions. It should be fixed now. Thank you for the sample data, it was really helpful.

**Comment by Dankomaister** [comment]
Thank you for fixing the problem so fast!
Will update schnetpack and continue training :)
